proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,367 @@
+concatMap under stream fusion.*  In turn, this could
+potentially make stream fusion general enough to replace foldr/build
+fusion in base.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal only changes the semantics of rewrite rules. No new syntax is introduced.
+
+Let us start with some terminology of rewrite rules.
+Consider the rule:
+::
+
+	{-# RULES "wombat"  forall f x.  foo x (\y. f y) = bar x f  #-}
+
+* 	**Template**.
+	The LHS of a rule is called its *template*.
+* 	**Template variables**.
+	The ``forall``'d variables are called the *template variables*.
+	In rule "wombat", ``f`` and ``x`` are template variables.
+* 	**Local binders**.
+	The *local binders* of a rule are the variables bound inside the template.
+	Example: ``y`` is a local binder of rule "wombat".
+	A local binder is specifically not a template variable, nor is it free in the entire rule.
+* 	**Target**.
+	The rule matcher matches the LHS of the rule (the template) against an expression in the program (the *target*).
+* 	**Substitution**.
+	A successful match finds a *substitution* S: a binding for each template variable, such that applying S to the LHS yields the target.
+* 	After a successful match we replace the target expression with the substitution S applied to the RHS of the rule.
+
+
+In GHC today, a template variable ``v`` matches any expression ``e`` if
+
+* ``e`` has the same type as ``v``
+* No local binder of the template is free in ``e``.
+
+The change proposed here makes matching more powerful by introducing the notion of a **higher order pattern**
+(a sub-expression of the template) that matches *any* target expression of the same type as the higher order pattern:
+
+* 	**Definition**.
+	A **higher order pattern (HOP)** is a sub-expression of the template of form ``f x y z`` where:
+
+	- ``f`` is a *template variable*
+	- ``x``, ``y``, ``z`` are *local binders* (like ``y`` in rule "wombat" above; see definitions).
+	- The arguments ``x``, ``y``, ``z`` are *distinct* variables
+	- ``x``, ``y``, ``z`` must be term variables (not type applications).
+
+* 	**Matching of higher order patterns (HOP-matching)**.
+        A higher order pattern ``f x y z`` (in the template) matches *any target expression* ``e`` provided:
+
+	- The target has the same type as the template
+	- No local binder is free in ``e``, other than ``x``, ``y``, ``z``.
+
+	If these two condition hold, the higher order pattern ``f x y z`` matches the target expression ``e``, yielding the substitution ``[f :-> \x y z. e]``.
+	Notice that this substitution is type preserving, and the RHS of the substitution has no free local binders.
+
+*       **Ambiguity breaking**.  When matching template ``(etmpl x)`` against target ``(etarget x)``, do *not* use HOP-matching even if ``(etmpl x)`` is a HOP; instead simply match ``etmpl`` (which is also a HOP) against ``etarget``.  See the next subsection for the justification for this refinement.
```

Comment:
This point isn't completely clear to me. I think it would be clearer to specify in general how a template application is matched against a target application, rather than merely giving a negative rule for when HOP-matching does not apply.

I think the rule is something like this.... When matching `(etmpl1 etmpl2)` against target `(etarget1 etarget2)`
 * If `(etmpl1 etmpl2)` is not a HOP, decompose the application (i.e. match `etmpl1` against `etarget1` and `etmpl2` against `etarget2`.
 * If `(etmpl1 etmpl2)` is a HOP, and `etarget2` has been renamed to be equal to `etmpl2` (which is necessarily a locally bound variable), decompose the application.
 * Otherwise, `(etmpl1 etmpl2)` is a HOP, so use HOP-matching against `(etarget1 etarget2)`.

Is that right?

Reply:
Thanks Adam.

@noughtmare I think that the "Ambiguity breaking" bullet is unnecessay.  There is no reason to *specify* that we must not use HOP-matching in the ambiguous case -- it's an implementation choice, since it's going to match either way.  And it's confusing to imply that it might matter.  I think we can simply drop the bullet altogether.

Then at the end of 3.1 instead of "The refined rule picks (2), which yields success with as few lambdas as possible in the match." we can simply say "It makes no difference which choice is made, so the implementation is free to choose either".

To return to Adam's question I think the *specificaiton* can be:
* If the template is a HOP, use HOP-matching
* If not, try to decompose applications

Simple.  We shouldn't over-specify.

@noughtmare ok?  I didn't want to make this change without seeing if you had some reason for the way it is.

Reply:
As the proposal says, HOP matches can cause eta-expansions compared to the status quo, which can have a performance impact. So, I still think the ambiguity breaking is a good idea.

Now that I'm thinking about it, there are also situations like matching the template `forall f. f x` with the target `foo x y x` (where `x` is local and `y` is not local), which I believe would not match due to ambiguity breaking. But there is a HOP match: `[f :-> \x -> foo x y x]`. 

So, it seems the ambiguity breaking rule does need to be modified, but I don't think removing it altogether is a good solution. Perhaps we can just say that HOP matches should be in maximally eta-reduced form? I believe that gets to the point without mentioning implementation details and it still seems relatively easy to implement.

Reply:
All good points.  Yes, I think we should modify

- If these two condition hold, the higher order pattern f x y z matches the target expression e, yielding the substitution [f :-> \x y z. e]. Notice that this substitution is type preserving, and the RHS of the substitution has no free local binders.

to say

- If these two condition hold, the higher order pattern f x y z matches the target expression e, yielding the substitution [f :-> e'], where e' is etaReduce( \x y z. e). Notice that this substitution is type preserving, and the RHS of the substitution has no free local binders.

Then we can add some words about `etaReduce`.   In the specification section we can just say "it's normal eta-reduction; e..g. `etaReduce( \x y z. f x y z) = f` and `etaReduce(\x y z.  f x y y z) = \x y. f x y y`.  In 3.1 we can motivate etaReduce.

OK?

Reply:
There is also of course the opposite, for example matching `forall f. f x` to `(\y -> foo y) x`, but I assume eta reduction is not so much of a problem?

Other than that I think your suggestion is good.

Reply:
> There is also of course the opposite, for example matching forall f. f x to (\y -> foo y) x, but I assume eta reduction is not so much of a problem?

Well, the pattern is a HOP, so HOP-matching would yield
```
  f :-> etaReduce(   \x.   (\y. foo y) x )
={do the eta-reduction}
  fo :-> \x.foo y
```
Just the same as non-HOP matching.  What's the problem here?

Reply:
First of all, a small typo is that it should be `f :-> \y. foo y`, but I thought you meant with `etaReduce` that it should do as much eta reduction as possible, so it would become `f :-> foo` in end, thus reducing too far.

If `etaReduce` only does one step then there is another problem if we have multiple HOP parameters, such as when we match `forall f. f x y` to `foo x y`, which should produce `f :-> foo`, but with only one eta reduction we will get `f :-> \x. foo x`.

Reply:
> so it would become f :-> foo in the end, thus reducing too far

Ah yes, how true.

What about:

"Matching of higher order patterns (HOP-matching). A higher order pattern `f x y z` (in the template) matches any target expression `e` provided:

1. The target has the same type as the template
2. No local binder is free in e, other than x, y, z.
3. The target `e` is not of form `(e' z)`, where `z` is not free in `e'`

If these three conditions hold,..."

Then we still need to explain (3), giving justification you give above.

Might you do that?  Joachim has merged the proposal but we can still make a new PR to clarify it.